### PR TITLE
Add support for strongly typed entitlements for usePlanshipCustomer hook

### PR DESCRIPTION
**Pull request by pwojnaro**
This PR, when merged, will add support for a strongly typed entitlements wrapper class when using `usePlanshipCustomer` hook. 

Example:

```
export class ClickerEntitlements extends EntitlementsBase {

  get subscriptionButtonClicks(): number {
    return this.entitlementsDict?.['subscription-button-clicks'].valueOf()
  }
  
  get buttonClicksPerMinute(): number {
    return this.entitlementsDict?.['button-clicks-per-minute'].valueOf()
  }
  
  get maxProjects(): number {
    return this.entitlementsDict?.['max-projects'].valueOf()
  }
  
  get premiumButton(): boolean {
    return this.entitlementsDict?.['premium-button'].valueOf()
  }
  
  get analyticsPanel(): boolean {
    return this.entitlementsDict?.['analytics-panel'].valueOf()
  }
  
  get projectTypes(): [] {
    return this.entitlementsDict?.['project-types'].valueOf()
  }
}


{ entitlements } = usePlanshipCustomer<ClickerEntitlements>(ClickerEntitlements)

// entitlements is of ClickerEntitlements type

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
...
 
   return (
     // Render a list of Planship customer subscriptions
   )
 }
 ```
 
+### Typed entitlements with PlanshipCustomer context provider
+
+When consuming PlanshipCustomer context in a TypeScript project, entitlements can be defined as a class with getters for individual levers, and passed to the `usePlanshipCustomer` hook.
```
Comment (trentpeterson) [inline README.md]: -> *the* PlanshipCustomer context

Remove comma after levers

File: README.md [hunk 2]
```diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
...
+}
+```
+
+This is helpful in certain IDEs like VS Code where it unlocks a full autocomplete experience for pricing entitlements entitlements.
+
+<img src="./static/entitlements-vscode.gif" alt="Animation showing entitlements autocomplete in VS Code">
+
 ## Using Planship providers in a Next.js app
 
 By default, all components in a Next.js app are React Server Components. React Server Components don't allow for consuming React context *directly*, so Planship context providers cannot by used *as is*. To address this, simply wrap them your own Client Component as outlined in [this official guide](https://vercel.com/guides/react-context-state-management-nextjs#rendering-third-party-context-providers-in-server-components).
```
Comment (trentpeterson) [inline README.md]: wrap them your own -> wrap them in your own

File: README.md [hunk 3]
```diff
@@ -198,11 +233,11 @@ export default async function RootLayout({
 }
 ```
 
-A complete working example of a Nuxt.js app integrated with Planship can be found at https://github.com/planship/planship-nextjs-demo
+This pattern can be explored in more detail in our Next.js app example app at https://github.com/planship/planship-nextjs-example
```
Comment (trentpeterson) [inline README.md]: Next.js app example app -> Next.js example app

File: src/usePlanshipCustomer.ts [hunk 4]
```diff
@@ -1,9 +1,26 @@
...
 }
 
-export { usePlanshipCustomer }
+export function usePlanshipCustomer(): IPlanshipCustomerContext
+export function usePlanshipCustomer<TEntititlements extends EntitlementsBase>(entitlementsType: {
+  new (entitlementsDict: Entitlements): TEntititlements
+}): ICustomerContext<TEntititlements>
+export function usePlanshipCustomer(entitlementsType?: { new (entitlementsDict: Entitlements): object }) {
+  const currentContext = useContext<IPlanshipCustomerContext>(context)
+  if (entitlementsType)
```
Comment (trentpeterson) [inline src/usePlanshipCustomer.ts]: Not sure if we're enforcing braces around if logic...

File: static/entitlements-vscode.gif
Comment (trentpeterson) [inline static/entitlements-vscode.gif:1]: Should consider running this through an app to shrink the size a bit. Nearly 400kb - could likely be sped up, frames removed, etc.

**Review by trentpeterson — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
...
+  // entitlements object is an instance of the Entitlements class
+  // with apiCallsPerMonth and advancedAnalytics properties
+
+  if (advancedAnalytics.advancedAnalytics) {
+    // do something if advancedAnalytics are enabled
+  }
+}
+```
+
+This is helpful in certain IDEs like VS Code where it unlocks a full autocomplete experience for pricing entitlements entitlements.
```
Comment (trentpeterson) [inline README.md]: Copy the language from the Vue module.

File: src/usePlanshipCustomer.ts (hunk 4 above)
Comment (pwojnaro) [reply src/usePlanshipCustomer.ts]: No, we do not. I just added it to eslint.rc, but I don't believe I enforce it in any other project (not a part of the recommended ruleset, seems arbitrary).

File: static/entitlements-vscode.gif
Comment (pwojnaro) [reply static/entitlements-vscode.gif:1]: I got it down to 220k

File: README.md [hunk 6]
```diff
@@ -148,7 +148,7 @@ export default function SubscriptionInfoComponent({ children }) {
 
 ### Typed entitlements with PlanshipCustomer context provider
 
-When consuming PlanshipCustomer context in a TypeScript project, entitlements can be defined as a class with getters for individual levers, and passed to the `usePlanshipCustomer` hook.
+When working with the entitlements dictionary returned by `usePlanshipCustomer` hook, it can be useful to wrap it in an object with getters for individual levers.
```
Comment (trentpeterson) [inline README.md]: returned by `usePlanshipCustomer` -> returned by the `usePlanshipCustomer`

**Review by trentpeterson — CHANGES_REQUESTED** [review]

File: src/usePlanshipCustomer.ts (hunk 4 above)
Comment (trentpeterson) [reply src/usePlanshipCustomer.ts]: Yeah, just thinking we should be consistent. No strong opinion but if the logic is on a separate line than the subsequent code it seems like braces help make it clearer.

**Review by pwojnaro — COMMENTED** [review]
Addressed additional feedback

**Review by trentpeterson — APPROVED** [review]
LGTM

**Pull request merged**